doc/user/content/ops/optimization.md

```diff
@@ -20,38 +20,84 @@ Building an efficient index for distinct **clauses** and **operators** can be pu
+| `WHERE x = 1 AND y = 'abc'`                       | `CREATE INDEX ON collection (x, y);`     |
+| `WHERE (x = 5 AND y = 'a') OR (x = 7 AND y = ''`) | `CREATE INDEX ON collection (x, y);`     |
+| `WHERE 2 * x = 64`                                | `CREATE INDEX ON collection (2 * x);`    |
+| `WHERE upper(y) = 'HELLO'`                        | `CREATE INDEX ON collection (upper(y));` |
 
-**Note:** to speed up a query using a multi-column index, as in `WHERE x = $1 AND y = $2`, the query must use all the fields in the index chained together via the `AND` operator.
+You can verify that Materialize is accessing the input by an index lookup using `EXPLAIN`. Check for `lookup` after the index name to confirm that an index lookup is happening, i.e., that Materialize is only reading the matching elements of the index instead of scanning the entire index:
+```
+materialize=> EXPLAIN SELECT * FROM foo WHERE x = 42 AND y = 'hello';
+                               Optimized Plan
+-----------------------------------------------------------------------------
+ Explained Query (fast path):                                               +
+   Project (#0, #1)                                                         +
+     ReadExistingIndex materialize.public.foo_x_y lookup value (42, "hello")+
+                                                                            +
+ Used Indexes:                                                              +
+   - materialize.public.foo_x_y                                             +
+```
+
+#### Matching multi-column indexes to multi-column `WHERE` clauses
+
+In general, your index key should exactly match the columns that are constrained in your `WHERE` clause. In more detail:
+- **Attention:** If your `WHERE` clause constrains fewer fields than your index key includes, then **the index will not be used**. For example, an index on `(x, y)` cannot be used to speed up `WHERE x = 7`.
+- If your `WHERE` clause constrains more fields than your index key includes, then the index might still provide some speedup, but it won't necessarily be optimal: In this case, the index lookup is performed using only those constraints that are included in the index key, and the rest of the constraints will be used to subsequently filter the result of the index lookup.
+- If you have an `OR` not all of whose arguments constrain the same fields, then create an index for the intersection of the constrained fields. For example, if you have `WHERE (x = 51 AND y = 'bbb') OR (x = 76 AND z = 9)`, then create an index just on `x`.
+- If you have an `OR` whose arguments constrain completely disjoint sets of fields (e.g., `WHERE x = 5 OR y = 'aaa'`), then try to rewrite your query using a `UNION` (or `UNION ALL`), where each argument of the `UNION` has one of the original `OR` arguments (with taking into account possible duplicate result rows from the intersection).
```

Comment:
The trouble is that blindly changing to either `UNION` or `UNION ALL` is wrong here:
1. If you use `UNION ALL`, then you will get extra duplicates for those records that pass both predicates (both arguments of the original OR).
2. If you use `UNION`, you lose duplicates everywhere.

It's mostly because of these complications that the optimizer is not doing this automatically at the moment (https://github.com/MaterializeInc/materialize/issues/4229). The user might be in a better better position to know whether 1. or 2. is appropriate for her use case:
1. might be good if the intersection is empty, or if duplicates are not an issue downstream.
2. might be good if there were no duplicates to begin with, or if losing duplicates is not an issue downstream.

So I think an extra warning here about duplicates is not redundant.

Reply:
Oh, what I meant is that I'd expect users to be aware of whether they need to use one or the other depending on whether they need to care or not about duplicates. But I understand your point, it definitely doesn't hurt!